### PR TITLE
Fix issues with incorrect serialization of LocalStorageDB items

### PR DESCRIPTION
**Pull request by pkukielka**
## Changes

To avoid further doubts I adjusted behaviour and interface of `LocalStorageDB`to be more in sunc with `vscode.Memento`, that is:
* setting key to `null` or `undefined` removes key from the store
* get on non-existing key returns an `undefined`

Note that it means setting key to `null` and then doing get on it will result in `undefined` value, not a `null`.

I also added exception handling in the `LocalStorageDB::get` to make sure we won't fail on previously incorrectly serialised values.

## Test plan

Unit tests which should cover all corner cases were added.

### CODE REVIEW

File: agent/src/global-state/AgentGlobalState.ts [hunk 1]
```diff
@@ -58,17 +58,17 @@ export class AgentGlobalState implements vscode.Memento {
         return [localStorage.LAST_USED_ENDPOINT, localStorage.ANONYMOUS_USER_ID_KEY, ...this.db.keys()]
     }
 
-    public get<T>(key: string, defaultValue?: unknown): any {
+    public get<T>(key: string, defaultValue?: T): T {
```
Comment (pkukielka) [inline agent/src/global-state/AgentGlobalState.ts:61]: Just so it have the same interface as `Memento` it inherits from.
That change is not really needed for that PR, so if someone disagrees I can revert.

File: agent/src/global-state/AgentGlobalState.ts [hunk 2]
```diff
@@ -123,12 +123,23 @@ class LocalStorageDB implements DB {
...
-    get(key: string): any {
+    get(key: string): any | undefined {
         const item = this.storage.getItem(key)
-        return item ? JSON.parse(item) : undefined
+        try {
+            return item ? JSON.parse(item) : undefined
+        } catch (error) {
+            // That should never happen now, but in past it was possible to store incorrectly serialized
+            // undefined values, which were failing to deserialize during the get operation
+            this.storage.removeItem(key)
```
Comment (dominiccooney) [inline agent/src/global-state/AgentGlobalState.ts:133]: WDYT about *not* removing these... we can keep running this `catch` branch. In the off chance there's some useful corrupt state...

File: agent/src/global-state/AgentGlobalState.ts [hunk 3]
```diff
@@ -123,12 +123,23 @@ class LocalStorageDB implements DB {
...
+            return undefined
+        }
     }
     set(key: string, value: any): void {
-        this.storage.setItem(key, JSON.stringify(value))
+        if (value) {
```
Comment (dominiccooney) [inline agent/src/global-state/AgentGlobalState.ts:138]: This might go too far... we're swapping `''` for `undefined` here now.

**Review by dominiccooney — CHANGES_REQUESTED** [review]
Could we do this with fewer semantic changes, to be conservative? Ideally we want something that lines up with whatever VSCode does, after you project it through our many layers of wrappers.

Some ideas but I have not carefully checked the details:

- Leave all the valid JSON values alone (including `''`, `null`.)
- If we treat the invalid JSON serialized string `"undefined"` as a missing key, then don't return that key in the keys enumeration. (If we treat it as `null` then having it in keys is probably fine...)
- Treat setting `undefined` as deleting, IIRC the client wants behavior like that.

File: agent/src/global-state/AgentGlobalState.ts (hunk 2 above)
Comment (pkukielka) [reply agent/src/global-state/AgentGlobalState.ts:133]: Sounds good, I have no problem with that.

**Comment by pkukielka** [comment]
> Ideally we want something that lines up with whatever VSCode does, after you project it through our many layers of wrappers.

> Leave all the valid JSON values alone (including `''`, `null`.)

We can pick one, but not both. 
I specifically did those semantic changes to make it very similar to VSCode, as I think that could reduce amount issues going forward. I do not think it is very risky either, as it only affects `null` / `undefined` values.

If we need to pick one of the above, I would rather go for correctness than tried to be conservative with changes, but possibly still wrong. But I'm open to other takes on it of course.

File: agent/src/global-state/AgentGlobalState.ts (hunk 3 above)
Comment (pkukielka) [reply agent/src/global-state/AgentGlobalState.ts:138]: Good point, I will add a test and fix it.

**Comment by dominiccooney** [comment]
Ok! Bringing it into line with VSCode sounds good to me.

**Review by dominiccooney — APPROVED** [review]
It is tempting to not `setItem(key, JSON.stringify(value))` when `value` is `undefined` but I guess this is fine, the `catch` will handle it on reading.

**Comment by pkukielka** [comment]
> It is tempting to not `setItem(key, JSON.stringify(value))` when `value` is `undefined` but I guess this is fine, the `catch` will handle it on reading.

Agree and fixed.
Problem was I did check `value != null && value != undefine` and then VSCode suggested me that I should replace it with value != null` because it test the same (which I did). But instead I should test for `value !== null && value !== undefine` which is clearly showing the intent (notice `!=` vs `!==`)

**Comment by dominiccooney** [comment]
> Problem was I did check value != null && value != undefine and then VSCode suggested me that I should replace it with value != null because it test the same (which I did).  ...

Ah good point, I forgot that `undefined == null` is `true`.

I like the explicit version anyway, thank you for updating it.
